Source/JavaScriptCore/runtime/WaiterListManager.cpp

```diff
@@ -50,38 +50,54 @@ WaiterListManager& WaiterListManager::singleton()
     return manager;
 }
 
-JSValue WaiterListManager::wait(JSGlobalObject* globalObject, VM& vm, void* ptr, AtomicsWaitValidation validation, Seconds timeout, AtomicsWaitType type)
+template <typename ValueType>
+JSValue WaiterListManager::waitImpl(JSGlobalObject* globalObject, VM& vm, ValueType* ptr, ValueType expectedValue, Seconds timeout, AtomicsWaitType type)
 {
     if (type == AtomicsWaitType::Async) {
         JSObject* object = constructEmptyObject(globalObject);
 
         bool isAsync = false;
         JSValue value;
-        if (validation == AtomicsWaitValidation::Fail)
-            value = vm.smallStrings.notEqualString();
-        else if (!timeout)
-            value = vm.smallStrings.timedOutString();
-        else {
-            isAsync = true;
-            JSPromise* promise = JSPromise::create(vm, globalObject->promiseStructure());
-            addAsyncWaiter(ptr, promise, timeout);
-            value = promise;
-        }
 
-        object->putDirect(vm, vm.propertyNames->async, jsBoolean(isAsync));
-        object->putDirect(vm, vm.propertyNames->value, value);
-        return object;
-    }
+        Ref<WaiterList> list = findOrCreateList(ptr);
+        {
+            Locker listLocker { list->lock };
+            if (WTF::atomicLoad(ptr) != expectedValue)
+                value = vm.smallStrings.notEqualString();
+            else if (!timeout)
+                value = vm.smallStrings.timedOutString();
+            else {
+                isAsync = true;
+                JSPromise* promise = JSPromise::create(vm, globalObject->promiseStructure());
+                Ref<Waiter> waiter = adoptRef(*new Waiter(promise));
```

Comment:
Do not perform GC allocation after taking a lock. It is unfortunate, but we should allocate promise before taking a lock.